Added ConfigOptions FlagSet

midnightconman opened the pull request:
Fixes: #11731 
Related to: #9701 

This PR creates a shared `config` flagset and converts all applications to use it. This will make it easier to ensure all applications behave the same way when using `config-path`, `job-config-path`, and `plugin-config-path`

Please let me know if you would like me to break this up, or if you guys can think of a better way to do config flag testing.

stevekuznetsov commented on prow/cmd/deck/main.go:
```diff
@@ -111,8 +112,10 @@ func (o *options) Validate() error {
 	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
-	fs.StringVar(&o.configPath, "config-path", "/etc/config/config.yaml", "Path to config.yaml.")
-	fs.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
+	for _, group := range []flagutil.OptionGroup{&o.config} {
```
You're missing kubernetes here

stevekuznetsov commented on prow/flagutil/config.go:
```diff
@@ -0,0 +1,74 @@
+
+// AddFlags injects config options into the given FlagSet.
+func (o *ConfigOptions) AddFlags(fs *flag.FlagSet) {
+	fs.BoolVar(&o.EnableConfig, "enable-config", true, "Enable config options, set false to disable config options.")
```
What is this?

midnightconman replied:
This was added to allow disabling of `os.Stat` checking during config validation. I welcome a better idea, this was what I came up with :smile:

cjwagner replied:
Whether or not we do validation should not be up to the user so I don't think this should be a flag. We may do different types of validation for different binaries, but the validation behavior should be fixed per binary, right?

midnightconman commented:
> We have some weird semantics:
> 
> * `plank` does not need to load plugin config
> * `plank` has no need to load the job config, only the main config
> * `hook` _must_ load plugin config
> * `checkconfig` _may_ load plugin config
> 
> There are oddities in a lot of our loading so I think we potentially need something less DRY -- see also the local tooling asking for `--dry-run` and a `deck` URL for an example of where we went too dry in the past.

I did notice the conversation in #prow slack today, thought that was a funny coincidence. I think we should add options and gates to validation, so we can disable `plugin-config-path` if that particular app doesn't need it. How does that sound?

> 
> Also, you forgot the kubernetes client flags in `deck`. Could we devise a test that validates that the same set of flags exists on binaries before and after this change? Otherwise it's hard to review. Adding or removing flags should be hard.

I'll fix that.

> 
> We might want to think about implementing the common lib first in a PR, then using one PR at a time to enable the use of that lib on other components.

I wouldn't have a problem with that, if you guys prefer it.

midnightconman commented:
/cc @fejta @cjwagner

midnightconman commented:
Whats the verdict here guys, should I break this up into smaller PRs? Should I throw out the `os.Stat` file validation, so testing is easier? Should I add additional gates to the validation function, so we can enable and disable validation of flag groups more intuitively?

Thank you for your time and reviews :smile:

cjwagner commented:
> should I break this up into smaller PRs?

Yes lets do what Steve suggested:

> We might want to think about implementing the common lib first in a PR, then using one PR at a time to enable the use of that lib on other components.

> Should I throw out the os.Stat file validation, so testing is easier?

It is unclear to me why this isn't just handled by the config agent. `config.Load` will fail if the file doesn't exist. A separate check seems unnecessary. It is also strange that this is made optional via `--enable-config` because the `Agent` function won't work if `--enable-config=false`.

> Should I add additional gates to the validation function, so we can enable and disable validation of flag groups more intuitively?

We shouldn't expose flags to gate this behavior as I noted above. The validation should be fixed and predictable per binary, I don't see any value in allowing users to disable validation with a flag. We do need to make the common validation library works for all of the following scenarios if that is what you mean though:

> * `plank` does not need to load plugin config
> * `plank` has no need to load the job config, only the main config
> * `hook` _must_ load plugin config
> * `checkconfig` _may_ load plugin config

midnightconman commented:
@stevekuznetsov @cjwagner @fejta - Please take another look.

This PR only contains `config.go` now, which only contains config related flagset. I also added options to enable string validation gates, so we can enable or disable validation from the individual binaries.

Thank you again for the reviews. :smile:

cjwagner commented on prow/flagutil/config.go:
```diff
@@ -0,0 +1,82 @@
+	// JobConfigPathNotEmpty is a boolean that enables validation of JobConfigPath string.
+	JobConfigPath bool
+	// PluginConfigPathNotEmpty is a boolean that enables validation of PluginConfigPath string.
+	PluginConfigPath bool
```
nit: This seems to me like we are shoehorning configs together. Plugin config is handled similarly to core config, but I don't think there is any benefit to combining it into this struct. I think having a `ConfigOptions` and a `PluginConfigOptions` might make more sense given that some things need config, some need plugin config, and some need both.
Similarly we could split `ConfigOptions` into a `CoreConfigOptions` and a `JobConfigOptions` (the latter containing a `CoreConfigOptions` since job config must be loaded with core config) so that binaries could use precisely the options that apply to them instead of a potentially partially initialized options struct.

midnightconman replied:
That works for me... take a look again please.

cjwagner commented on prow/flagutil/config.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// Agent returns a started config agent.
+func (o *ConfigOptions) Agent() (agent *config.Agent, err error) {
```
We would need a parallel to this function for the plugin config agent as well.

midnightconman replied:
Done.

midnightconman commented:
@stevekuznetsov @cjwagner @fejta - Please take another look.

I split out `CoreConfigOptions` and `PluginConfigOptions`, like @cjwagner requested.

stevekuznetsov commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,58 @@
+	if err != nil {
+		return nil, err
+	}
+	agent.Set(config)
```
I don't think we need to do the `Load` and `Set` -- the first call to `Start` will handle that for us

stevekuznetsov replied:
(and `Load` calls `Set` for plugin config anyway)

midnightconman replied:
Can you be a little more specific here, please? I think we need to just do a `config.Load()`, right?

stevekuznetsov commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,58 @@
+// AddFlags injects config options into the given FlagSet.
+func (o *CoreConfigOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.ConfigPath, "config-path", "/etc/config/config.yaml", "Path to config.yaml.")
+	fs.StringVar(&o.JobConfigPath, "job-config-path", "", "Path to prow job configs.")
```
How will this support `plank` or other components that want the `config-path` but not the `job-config-path`?

midnightconman replied:
This change leaves validation to the individual binaries, so plank would just not validate anything for `job-config-path`. Then the default would be set `""`.

stevekuznetsov replied:
But we don't want to expose flags that are not supposed to be used or set, though, right? That was one of the mistakes with the prior DRY flag approaches -- end up with `-deck-url` flag on all sorts of binaries that could never use it

midnightconman replied:
Okie dokie, splitting it out.

stevekuznetsov commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// Validate validates config options.
+func (o *CoreConfigOptions) Validate(dryRun bool) error {
```
Are we bound to use the `dryRun` here? It is not relevant, right?

midnightconman replied:
Nope, I'll remove it.

midnightconman commented:
@stevekuznetsov @cjwagner @fejta - Please take another look.

I split out `core_config` and `job_config`. Let me know if you think there is a better way to do Agent for `job_config`, not sure we wan't to have a separate agent. Also, removed `agent.Set()`.

stevekuznetsov commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+// Agent returns a started config agent.
+func (o *CoreConfigOptions) Agent() (agent *config.Agent, err error) {
+	agent = &config.Agent{}
+	config, err := config.Load(o.ConfigPath, "")
```
Sorry, I meant that we don't need the `Load()` _or_ the `Set()` calls -- the impl for `Start()` for both agents does a pre-emptive `Load()` by itself

stevekuznetsov commented on prow/flagutil/job_config.go:
```diff
@@ -0,0 +1,56 @@
+
+// AddFlags injects job config options into the given FlagSet.
+func (o *JobConfigOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.JobConfigPath, "job-config-path", "", "Path to prow job configs.")
```
Do we have a use-case for only loading job config and not core config? `hook`?

Also, since many will need both -- can we implement something (maybe using struct embedding) that gives downstream callers a simple way to request both flags without having downstream code to funnel the configPath from one place to another? Hopefully we can make the actual paths opaque to the callers (and unexported)

stevekuznetsov commented on prow/flagutil/job_config.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Validate validates job config options.
+func (o *JobConfigOptions) Validate() error {
```
Since this is now specifically opt-in we can validate that something was set

stevekuznetsov commented on prow/flagutil/plugin_config.go:
```diff
@@ -0,0 +1,56 @@
+// Agent returns a started plugin agent.
+func (o *PluginConfigOptions) Agent() (agent *plugins.Agent, err error) {
+	agent = &plugins.Agent{}
+	config, err := plugins.Load(o.PluginConfigPath)
```
Similar here, we can just do the call to `Start()`

fejta commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+// CoreConfigOptions holds options for interacting with Configs.
+type CoreConfigOptions struct {
```
`type CoreConfigOptions string`

What is the need for a struct? Same for the others.

fejta commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+
+// AddFlags injects config options into the given FlagSet.
+func (o *CoreConfigOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.ConfigPath, "config-path", "/etc/config/config.yaml", "Path to config.yaml.")
```
If we are going to refactor --config-path, please make sure the default value becomes "" rather than tightly coupling to a particular volume mount.

fejta commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+// Validate validates config options.
+func (o *CoreConfigOptions) Validate() error {
```
Please drop this function.

fejta commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+// Agent returns a started config agent.
+func (o *CoreConfigOptions) Agent() (agent *config.Agent, err error) {
```
Please do not use named return parameters.

fejta commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+
+// Agent returns a started config agent.
+func (o *CoreConfigOptions) Agent() (agent *config.Agent, err error) {
+	agent = &config.Agent{}
```
Don't allocate this memory until we know the config is good

fejta commented on prow/flagutil/core_config.go:
```diff
@@ -0,0 +1,55 @@
+		return nil, err
+	}
+
+	err = agent.Start(o.ConfigPath, "")
```
This whole function seems more appropriate as a function in the config package:

```golang
package config
func NewAgent(corePath, jobPath string) (*config.Agent, error) {
  cfg, err := config.Load(corepath, jobPath)
  var agent config.Agent
  err = agent.Start(corePath, jobPath)
  return &agent, nil
}

stevekuznetsov replied:
Yeah, maybe _this_ is the DRY we really need? Flag _names_ being identical is nice, maybe we could have constants for those .. the rest .. yeah it's hard to DRY since everyone does it differently
